File: src/server/lib/version-persistence.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { writeFileAtomic } from "./atomic-write.js";

const VERSION_FILE_NAME = "claude-version.json";

export function readPersistedVersion(dataDir: string): string | null {
  const filePath = join(dataDir, VERSION_FILE_NAME);
  if (!existsSync(filePath)) return null;
  try {
    const raw: unknown = JSON.parse(readFileSync(filePath, "utf-8"));
    if (
      typeof raw === "object" &&
      raw !== null &&
      "version" in raw &&
      typeof (raw as Record<string, unknown>).version === "string"
    ) {
      return (raw as Record<string, unknown>).version as string;
    }
    return null;
  } catch {
    return null;
  }
}

export function writePersistedVersion(dataDir: string, version: string | null): void {
  const filePath = join(dataDir, VERSION_FILE_NAME);
  writeFileAtomic(filePath, `${JSON.stringify({ version }, null, 2)}\n`);
}

File: server/lib/versionpersistence.go
package lib

import (
	"encoding/json"
	"os"
	"path/filepath"
)

const versionFileName = "claude-version.json"

// versionFile is the JSON structure of the version persistence file.
type versionFile struct {
	Version *string `json:"version"`
}

// ReadPersistedVersion reads the persisted version from dataDir/claude-version.json.
// Returns "" on any error (missing file, malformed JSON, missing or non-string version field).
func ReadPersistedVersion(dataDir string) string {
	filePath := filepath.Join(dataDir, versionFileName)
	data, err := os.ReadFile(filePath)
	if err != nil {
		return ""
	}
	var v versionFile
	if err := json.Unmarshal(data, &v); err != nil {
		return ""
	}
	if v.Version == nil {
		return ""
	}
	return *v.Version
}

// WritePersistedVersion writes the given version string to dataDir/claude-version.json.
// Creates dataDir if it does not exist. Uses WriteFileAtomic for safe writes.
// An empty version string is stored as {"version":null}.
func WritePersistedVersion(dataDir string, version string) error {
	if err := os.MkdirAll(dataDir, 0o755); err != nil {
		return err
	}
	filePath := filepath.Join(dataDir, versionFileName)

	var v versionFile
	if version != "" {
		v.Version = &version
	}
	// v.Version remains nil for empty string → marshals as {"version":null}

	out, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	out = append(out, '\n')
	return WriteFileAtomic(filePath, out)
}
